core/src/main/kotlin/org/evomaster/core/problem/rest/classifier/probabilistic/knn/KNN400EndpointModel.kt

```diff
@@ -0,0 +1,115 @@
+): AbstractProbabilistic400EndpointModel(endpoint, warmup, dimension, encoderType, randomness) {
+
+    private val samples = mutableListOf<Pair<List<Double>, Int>>()  // (features, statusCode)
+
+    fun getSamples(): List<Pair<List<Double>, Int>> = samples
+
+    // Euclidean distance between two points in the feature space
+    private fun distance(a: List<Double>, b: List<Double>): Double {
+        return sqrt(a.zip(b).sumOf { (ai, bi) -> (ai - bi) * (ai - bi) })
+    }
+
+    override fun classify(input: RestCallAction): AIResponseClassification {
+        verifyEndpoint(input.endpoint)
+
+        // treat empty action as "unknown", avoid touching the model
+        if (input.parameters.isEmpty()) {
+            return AIResponseClassification()
+        }
+
+        val encoder = InputEncoderUtilWrapper(input, encoderType = encoderType)
+        val inputVector = encoder.encode()
+
+        initializeIfNeeded(inputVector)
+
+        if (getModelAccuracyFullHistory().totalSentRequests < warmup) {
+            // Return equal probabilities during warmup
+            return AIResponseClassification(
+                probabilities = mapOf(
+                    200 to 0.5,
+                    400 to 0.5
+                )
+            )
+        }
+
+        if (inputVector.size != dimension) {
+            throw IllegalArgumentException("Expected input vector of size ${this.dimension} but got ${inputVector.size}")
+        }
+
+        // Find k nearest neighbors
+        val neighbors = samples
+            .map { (x, label) -> distance(inputVector, x) to label }
+            .sortedBy { it.first }
+            .take(k)
+
+        // Majority vote
+        val votes = neighbors.groupingBy { it.second }.eachCount()
+
+        // Convert to pseudo probabilities (votes / k)
+        val probabilities = votes.mapValues { it.value.toDouble() / k }
+
+        return AIResponseClassification(probabilities = probabilities)
+    }
+
+    override fun updateModel(input: RestCallAction, output: RestCallResult) {
+
+        verifyEndpoint(input.endpoint)
+
+        // Ignore empty action
+        if (input.parameters.isEmpty()) {
+            return
+        }
+
+        val encoder = InputEncoderUtilWrapper(input, encoderType = encoderType)
+        var inputVector = encoder.encode()
+
+        initializeIfNeeded(inputVector)
+
+        if (inputVector.size != this.dimension) {
+            throw IllegalArgumentException("Expected input vector of size ${this.dimension} but got ${inputVector.size}")
+        }
+
+        /**
+         * Updating classifier performance based on its prediction
+         */
+        val trueStatusCode = output.getStatusCode()
+        updatePerformance(input, trueStatusCode)
+
+        /**
+         * Store only classes of interest (i.e., 400 and not 400 groups)
+         */
+        if (trueStatusCode == 400) {
+            samples.add(inputVector to 400)
+        } else {
+            samples.add(inputVector to 200)
```

Comment:
a 200 here might look confusing. could have a constant `NOT_400 = 200` at class level, and use it here for readability